File: github_code_index_sort/cpp/4/33.c
//	SortingProject.cpp
//	Joshua Castelli

#include "stdafx.h"
#include "heapSort.h"
#include <math.h>
#include <time.h>

int heapSort::parent(int i) { return (floor(i/2)); };
int heapSort::leftChild(int i) { return (2 * i); };
int heapSort::rightChild(int i) { return ((2*i) + 1); };
	heapSort::heapSort() {};
void heapSort::sort(string* list, int size) {
		string* heap = list;
		int heapSize = size-1;
		sortHeap(heap, heapSize);
	};
void heapSort::maxHeapify(string* heap, int i, int heapSize) {

	left = leftChild(i);
	right = rightChild(i);

	if (left <= heapSize && heap[left] > heap[right]) {
		largest = left;
	}
	else {
		largest = i;
	}

	if (right <= heapSize && heap[right] > heap[largest]) {
		largest = right;
	}

	if (largest != i) {
		string temp = heap[i];
		heap[i] = heap[largest];
		heap[largest] = temp;
		maxHeapify(heap, largest, heapSize);
	}

};
void heapSort::buildMaxHeap(string* heap, int heapSize) {
	clock_t timerStart;
	timerStart = clock();

	for (int i = (floor(heapSize / 2)); i >= 0; i--) {
		maxHeapify(heap, i, heapSize);
	}

	cout << "BuildHeap Time: " << ((float)(clock() - timerStart) / CLOCKS_PER_SEC) * 1000 << "ms" << endl;
};
void heapSort::sortHeap(string* heap, int heapSize) {
	buildMaxHeap(heap, heapSize);
	for (int i = heapSize; i >= 1; i--) {
		string temp = heap[0];
		heap[0] = heap[i];
		heap[i] = temp;
		heapSize--;
		maxHeapify(heap,0,heapSize);
	}
};
void heapSort::maxHeapInsert() {
	//not implemented yet.
};
string heapSort::heapExtracMax(string* heap, int heapSize) {
	string max;
	if (heapSize < 0) {
		cout << "error, heap underflow" << endl;
	}
	else {
		max = heap[0];
		heap[0] = heap[heapSize];
		heapSize--;
		maxHeapify(heap,0,heapSize);
	}
	return max;
};
void heapSort::heapIncreaseKey() {
	//not implemented yet.
};
string heapSort::heapMaximum(string* heap) { return heap[0]; };

File: github_code_index_sort/java/4/33.java
public class HeapSortJava {

    public void heapify(int arr[], int n, int root) {
        int largestNumber = root;
        int left = 2 * root + 1;
        int right = 2 * root + 2;

        if (left < n && arr[left] > arr[largestNumber]) {
            largestNumber = left;
        }

        if (right < n && arr[right] > arr[largestNumber]) {
            largestNumber = right;
        }

        if (largestNumber != root) {
            int swap = arr[root];
            arr[root] = arr[largestNumber];
            arr[largestNumber] = swap;
            heapify(arr, n,largestNumber);
        }
    }

    public void sort(int arr[]) {
        int len = arr.length;

        for (int i = n / 2 - 1; i >= 0; i--) {
            heapify(arr, n, i);
        }

        for (int i = n - 1; i >= 0; i--) {
            int temp = arr[0];
            arr[0] = arr[i];
            arr[i] = temp;

            heapify(arr, i, 0);
        }
    }

    static void printArray(int arr[]) {
        int len = arr.length;
        for (int i = 0; i < len; i++) {
            System.out.println(arr[i] + " ");
        }
        System.out.println();
    }

    public static void main(String args[]) {
        int arr[] = {34, 33, 75, 44, 89, 01};
        int len = arr.length;

        HeapSort heapSort = new HeapSort();
        heapSort.sort(arr);

        System.out.println("Sorted array is ");
        printArray(arr);
    }
}